modules/universalId.js

```diff
@@ -0,0 +1,406 @@
+ */
+export function getValidSubmoduleConfigs(allSubmoduleConfigs, submodules) {
+  if (!Array.isArray(allSubmoduleConfigs)) {
+    return [];
+  }
+
+  // get all enabled storage types to validate submoduleConfig.storage.type
+  const storageTypes = [];
+  if (browserSupportsLocalStorage()) {
+    storageTypes.push(LOCAL_STORAGE);
+  }
+  if (utils.cookiesAreEnabled()) {
+    storageTypes.push(COOKIE);
+  }
+
+  // get all submodule names to validate submoduleConfig.name
+  const submoduleNames = submodules.reduce((carry, submodule) => {
+    carry.push(submodule.name);
+    return carry;
+  }, []);
+
+  return allSubmoduleConfigs.reduce((carry, submoduleConfig) => {
+    if (!submoduleConfig || typeof submoduleConfig !== 'object' ||
+      typeof submoduleConfig.name !== 'string' || submoduleConfig.name.length === 0 || submoduleNames.indexOf(submoduleConfig.name) === -1) {
+      // invalid
+      return carry;
+    }
+
+    if (submoduleConfig.storage && typeof submoduleConfig.storage === 'object' &&
+      typeof submoduleConfig.storage.type === 'string' && submoduleConfig.storage.type.length &&
+      typeof submoduleConfig.storage.name === 'string' && submoduleConfig.storage.name.length &&
+      storageTypes.indexOf(submoduleConfig.storage.type) !== -1) {
+      // valid
+      carry.push(submoduleConfig);
+    } else if (submoduleConfig.value && typeof submoduleConfig.value === 'object' && Object.keys(submoduleConfig.value).length) {
+      // valid
+      carry.push(submoduleConfig);
+    }
+    return carry;
+  }, []);
+}
+
+/**
+ * @param {IdSubmodule[]} allSubmodules
+ * @param {IdSubmoduleConfig[]} validSubmoduleConfigs
+ * @returns {IdSubmodule[]}
+ */
+export function getValidSubmodules(allSubmodules, validSubmoduleConfigs) {
+  if (validSubmoduleConfigs.length === 0) {
+    return [];
+  }
+  return allSubmodules.reduce((carry, submodule) => {
+    if (find(validSubmoduleConfigs, submoduleConfig => submoduleConfig.name === submodule.name)) {
+      // valid
+      carry.push(submodule);
+    }
+    return carry;
+  }, []);
+}
+
+/**
+ * @param config
+ * @param {IdSubmodule[]} allSubmodules
+ */
+export function init (config, allSubmodules) {
+  syncDelay = 0;
+  submodules = [];
+  extendBidData = [];
+  submoduleConfigs = [];
+
+  if (utils.cookiesAreEnabled()) {
+    if (document.cookie.indexOf(OPT_OUT_COOKIE) !== -1) {
+      utils.logInfo(`${UNIVERSAL_ID} - opt-out cookie found, exit module`);
+      return;
+    }
+  }
+  // listen for config userSyncs to be set
+  config.getConfig('usersync', ({usersync}) => {
+    if (usersync) {
+      utils.logInfo(`${UNIVERSAL_ID} - usersync config updated`);
+
+      syncDelay = usersync.syncDelay || 0;
+      submoduleConfigs = getValidSubmoduleConfigs(usersync.universalIds, allSubmodules);
+      submodules = getValidSubmodules(allSubmodules, submoduleConfigs);
+      if (submodules.length) {
+        $$PREBID_GLOBAL$$.requestBids.addHook(requestBidHook);
```

Comment:
If this module is meant to be ran after `consentManagement` and `pubCommonId` (to ensure their information can be collected), we may need to specify a priority here when adding the hook.  

Below is list of the current hooks on this function and their priorities:
- `consentManagement` currently has a priority value of `50`
- `pubCommonId` currently doesn't have a priority defined - so it's using the default priority of `10` (same as this module).  
- `pre1api` currently has a priority value of `5`

Per my understanding, there can be two hooks at the same priority on the same function; they'll just execute in the order whichever was added first.  But if we want to guarantee the sequencing of modules here, we should tweak the priority either here or in `pubCommonId` (I'd be leaning on the latter).

Reply:
that's correct